### ISSUE TITLE
[RFC] Harmonisation of I/O abstractions in stdlib

### ISSUE THREAD

**Issue by zielmicha** [nim-lang/Nim#5728]
I think that it's high time to tidy and improve I/O abstractions in stdlib. Currently we have:
- `system.File`, which is just `FILE*` from libc
- `streams.Stream` which is synchronous Input/Output stream with ad-hoc runtime polymorphism (with implementations for `osproc` and `system.File`)
- `net.Socket` which is synchronous socket interleaved with SSL implementation
- `asyncnet.AsyncSocket` which is asynchronous socket interleaved with SSL implementation

I think this is too messy and needs improvements. (That's one of the main reasons I've started [reactor.nim](https://github.com/zielmicha/reactor.nim)).

Summary of my improvements ideas:
- adding `View` type, which is a struct holding pointer+size (similar to Go slice) (Nim strings are really hard too use when avoiding copies is important)
- `ByteInput`, `ByteOutput` and `ByteStream` concepts with `read/write` and `close` methods (of course, exact concept names are not important), which take `View` as parameter
- improved `Future` without allocations on non-blocking operations (as in [reactor.nim Future](https://github.com/zielmicha/reactor.nim/blob/master/reactor/async/future.nim))
- `AsyncByteInput`, `AsyncByteOutput` and `AsyncByteStream` concepts similar to `ByteInput`/`ByteOutput`/`ByteStream`, but returning `Future`
- `sockets.Socket`, `osproc` and `system.File` implementing `ByteStream`
- SSL support separated from socket implementation, acting on `[Async]ByteStream`

I think most of this can be done without affecting backwards compatibility too much. I would really like Nim to have clean and efficient I/O library - either by improving standard library or continuing work on reactor.nim.

Some of these changes would be much easier with builtin runtime polymorphism (aka interfaces). Recently there was a rather old commit by @zah merged that added mention of vtptr/vtref to the manual. Is vtref going to be implemented in Nim? 

I'd like to hear from other Nim developers, if you feel that changes to I/O portion of standard library are needed and what's you tolerance on backward incompatible changes. And of course if my ideas are good or you have different take on solving the issue.
If the overall response is positive, I'm going to create more RFCs describing my proposal in detail.

**Comment by Araq** [comment]
I agree with pretty much everything you say. I like how streams.nim does polymorphism though. It's hard to do it more efficiently (try it! ;-) ) and if anything could be abstracted by a macro that produces the same code. Even the planned interface feature that concepts give us won't match its speed as it means we pass fat pointers around instead.

**Comment by yglukhov** [comment]
I haven't found any mentions regarding dispatcher interface. It would be good to figure out an abstract interface for dispatcher to allow custom dispatchers. E.g. CFRunloop on mac/ios, ALooper on Android.

**Comment by zah** [comment]
I also think we need the dispatcher defined as a concept. One unusual implementation I care about is providing a dispatcher based on the internal async APIs of Nginx, which would allow for very efficient Nim web apps, implemented as Nginx dynamic modules.
